packages/kit/src/runtime/server/page/actions.js

```diff
@@ -0,0 +1,153 @@
+import { error, json } from '../../../exports/index.js';
+import { normalize_error } from '../../../utils/error.js';
+import { negotiate } from '../../../utils/http.js';
+import { HttpError, Redirect, ValidationError } from '../../control.js';
+import { error_to_pojo } from '../utils.js';
+
+// Info: `enhance` action and `updateForm` live in `runtime/app/forms.js`
+
+/** @param {import('types').RequestEvent} event */
+export function is_action_json_request(event) {
+	const accept = negotiate(event.request.headers.get('accept') || 'text/html', [
+		'text/html',
+		'application/json'
+	]);
+
+	return (
+		accept === 'application/json' &&
+		event.request.method !== 'GET' &&
+		event.request.method !== 'HEAD'
+	);
+}
+
+/**
+ * @param {import('types').RequestEvent} event
+ * @param {import('types').SSROptions} options
+ * @param {import('types').SSRNode['server']} server
+ */
+export async function handle_action_json_request(event, options, server) {
+	const actions = server.actions;
+
+	if (!actions) {
+		maybe_throw_migration_error(server);
+		// TODO should this be a different error altogether?
+		return new Response('POST method not allowed. No actions exist for this page', {
+			status: 405,
+			headers: {
+				// https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/405
+				// "The server must generate an Allow header field in a 405 status code response"
+				allow: 'GET'
+			}
+		});
+	}
+
+	try {
```

Comment:
I get that the "success and stay on page" case and the "invalid form" case are similar from the framework standpoint (i.e., in this try/catch.) Thus, `return`ing the error rather than `throw`ing. But it's inconvenient on the userland side. It makes it less easy to split out logic into functions that `return` something on success and `throw` errors.  

For example, I'd really rather do this...

```ts
/**
 * @throws ValidationError 
 */
const validateSignInForm = (raw: Record<string, unknown>): SignInFormData => {
  // perform basic validation, return the shape
}
/**
 * @throws ValidationError 
 */
 const signIn = async (event: RequestEvent, data: SignInFormData): Promise<LoginResult> => {
  // check email and password in the db, read/set cookies, return result
}
export const actions = {
  default: async (event: RequestEvent) => {
    const data =  validateSignInForm(myFormDataToPojo(await event.request.formData()));
    const result = await signIn(event, data);
    setFlashMessage(event, `Welcome, ${result.user.profile.name}.`);
    return redirect(303, result.authRedirect)
  }
}

```
...than this (using the same helpers)...

```ts
export const actions = {
  default: async (event: RequestEvent) => {
    let data: SignInFormData;
    try {
      data =  validateSignInForm(myFormDataToPojo(await event.request.formData()));
    } catch (error) {
      if (error instanceof ValidationError) {
        return error
      }
      throw error;
    }
    let result: LoginResult
    try {
      result = await signIn(event, data);
      
      
    } catch (error) {
      if (error instanceof ValidationError) {
        return error;
      }
      throw error;
    }

    // I think this has to be here since we can't put the throw in the second catch above...
    setFlashMessage(event, `Welcome, ${result.user.profile.name}.`);
    throw redirect(303, result.authRedirect)
   
  }
}
```

On the face of it this can be fixed pretty easily here. I know there was some concern about type safety if ValidationError is thrown rather than returned, but I think it's worth figuring that out separately rather than making the developer jump. through hoops and going against the convention.

Reply:
The types can't be infered if you `throw invalid`, which means your type safety would be lost. Only infering them for the success case goes only half-way. I can understand the desire to `throw` instead though. There's not perfect solution that makes all sides happy, unfortunately.

Reply:
We don't need to make `ValidationError` importable — it's all in userland. As long as both the action and the `validateSignInForm` and `signIn` functions are all referencing the same thing, it's fine. Just need to do this:

```diff
} catch (e) {
-  if (e instanceof ValidationError) {
-    return e;
+  if (e instanceof MyValidationError) {
+    return invalid(e.status, e.data);
  }
  throw e;
}
```

If this is something you need to do in multiple actions, it can easily be abstracted out into a function.

This might seem inconvenient in the example above, but there are other patterns that don't involve try-catch:

```js
const [errors, data] = validateSignInForm(myFormDataToPojo(await event.request.formData()));
if (errors) return invalid(400, errors);
```

In general we don't want to encourage overuse of `throw` as a control flow mechanism — it makes code harder to follow. Returning validation errors does make more sense from the framework standpoint, but I would also argue it makes more sense from the app standpoint — rather than having three types of errors...

* `throw error(...)` causes the `+error` page to render
* `throw new Error(...)` causes the `+error` page to render, but also triggers `handleError`
* `throw invalid(...)` renders the same page as would have been rendered if nothing was thrown

...we have a simple rule that applies to `load` and `actions` and handlers in `+server.js` — if you `throw` (either an error or a redirect), the page will not be rendered; if you `return`, it will. This is a case where I'd strongly argue for clarity over convenience.

Reply:
Ok. Got it. Just my ingrained mantra of "throw error, return success." I can deal with it, either as is or by making a wrapper. Thanks.